File: frontend/src/api/modules.ts
import { API_BASE_URL, getErrorMessage } from "./rules";

export type ModuleSummary = {
  id: string;
  title: string;
  system: string;
  language: string;
  keeper_notes?: string | null;
  spoiler_level: string;
  has_documents: boolean;
  document_count: number;
  index_ready: boolean;
};

export type ModuleCreateRequest = {
  id: string;
  title: string;
  system?: string;
  language?: string;
  keeper_notes?: string;
  spoiler_level?: string;
};

export type ModuleSource = {
  knowledge_base: string;
  module_id: string;
  source: string;
  title_path: string[];
  content_type: string;
  spoiler_level: string;
  score?: number;
  content_preview: string;
};

export type ModuleAskRequest = {
  question: string;
  top_k?: number;
  rerank_top_k?: number;
};

export type ModuleAskResponse = {
  answer: string;
  sources: ModuleSource[];
};

export type ModuleIngestResponse = {
  module_id: string;
  document_count: number;
  chunk_count: number;
  index_dir: string;
};

export type ModuleStructureItem = {
  label: string;
  source: string;
  title_path: string[];
  content_type: string;
  preview: string;
};

export type ModuleStructureGroup = {
  content_type: string;
  label: string;
  items: ModuleStructureItem[];
};

export type ModuleStructureResponse = {
  module_id: string;
  groups: ModuleStructureGroup[];
};

export async function listModules(): Promise<ModuleSummary[]> {
  const response = await fetch(`${API_BASE_URL}/api/v1/modules`);
  if (!response.ok) {
    throw new Error(await getErrorMessage(response));
  }
  return response.json();
}

export async function createModule(request: ModuleCreateRequest): Promise<ModuleSummary> {
  const response = await fetch(`${API_BASE_URL}/api/v1/modules`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(request)
  });
  if (!response.ok) {
    throw new Error(await getErrorMessage(response));
  }
  return response.json();
}

export async function uploadModuleDocument(moduleId: string, file: File): Promise<void> {
  const formData = new FormData();
  formData.append("file", file);
  const response = await fetch(`${API_BASE_URL}/api/v1/modules/${moduleId}/upload`, {
    method: "POST",
    body: formData
  });
  if (!response.ok) {
    throw new Error(await getErrorMessage(response));
  }
}

export async function ingestModule(moduleId: string): Promise<ModuleIngestResponse> {
  const response = await fetch(`${API_BASE_URL}/api/v1/modules/${moduleId}/ingest`, {
    method: "POST"
  });
  if (!response.ok) {
    throw new Error(await getErrorMessage(response));
  }
  return response.json();
}

export async function askModule(moduleId: string, request: ModuleAskRequest): Promise<ModuleAskResponse> {
  const response = await fetch(`${API_BASE_URL}/api/v1/modules/${moduleId}/ask`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(request)
  });
  if (!response.ok) {
    throw new Error(await getErrorMessage(response));
  }
  return response.json();
}

export async function getModuleStructure(moduleId: string): Promise<ModuleStructureResponse> {
  const response = await fetch(`${API_BASE_URL}/api/v1/modules/${moduleId}/structure`);
  if (!response.ok) {
    throw new Error(await getErrorMessage(response));
  }
  return response.json();
}

File: backend/app/schemas/modules.py
from pydantic import BaseModel, Field


class ModuleMetadata(BaseModel):
    id: str = Field(min_length=1, pattern=r"^[a-z0-9_-]+$")
    title: str = Field(min_length=1)
    system: str = Field(default="coc7")
    language: str = Field(default="zh")
    keeper_notes: str | None = None
    spoiler_level: str = Field(default="keeper_only")


class ModuleCreateRequest(ModuleMetadata):
    pass


class ModuleSummary(ModuleMetadata):
    has_documents: bool
    document_count: int
    index_ready: bool


class ModuleDetail(ModuleSummary):
    module_dir: str
    documents_dir: str
    processed_dir: str
    index_dir: str


class ModuleAskRequest(BaseModel):
    question: str = Field(min_length=1)
    top_k: int | None = Field(default=None, ge=1, le=20)
    rerank_top_k: int | None = Field(default=None, ge=1, le=10)


class ModuleSource(BaseModel):
    knowledge_base: str = "module"
    module_id: str
    source: str
    title_path: list[str] = Field(default_factory=list)
    content_type: str = "document"
    spoiler_level: str = "keeper_only"
    score: float | None = None
    content_preview: str


class ModuleAskResponse(BaseModel):
    answer: str
    sources: list[ModuleSource]


class ModuleIngestResponse(BaseModel):
    module_id: str
    document_count: int
    chunk_count: int
    index_dir: str


class ModuleUploadResponse(BaseModel):
    module_id: str
    filename: str
    path: str
